dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.cpp

```diff
@@ -183,12 +199,15 @@ Block ColumnFileTiny::readBlockForMinorCompaction(const PageReader & page_reader
     }
 }
 
-ColumnTinyFilePtr ColumnFileTiny::writeColumnFile(DMContext & context, const Block & block, size_t offset, size_t limit, WriteBatches & wbs, const BlockPtr & schema, const CachePtr & cache)
+ColumnTinyFilePtr ColumnFileTiny::writeColumnFile(DMContext & context, const Block & block, size_t offset, size_t limit, WriteBatches & wbs, const CachePtr & cache)
 {
     auto page_id = writeColumnFileData(context, block, offset, limit, wbs);
-    auto new_column_file_schema = schema ? schema : std::make_shared<Block>(block.cloneEmpty());
+
+    auto new_digest = hashSchema(block);
+    auto schema = getSharedBlockSchemas(context)->getOrCreate(new_digest, block);
+
```

Comment:
Ditto. Maybe you can just `auto schema = getSharedBlockSchemas(context)->getOrCreate(block)`.